fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/DefaultResourcePackMixin.java

```diff
@@ -19,55 +19,110 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.JarURLConnection;
 import java.net.URL;
-import java.util.Enumeration;
+import java.net.URLConnection;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.function.Predicate;
 
+import org.jetbrains.annotations.Nullable;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.injection.At;
-import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.Overwrite;
+import org.spongepowered.asm.mixin.Unique;
 
 import net.minecraft.resource.DefaultResourcePack;
 import net.minecraft.resource.DirectoryResourcePack;
+import net.minecraft.resource.ResourcePack;
 import net.minecraft.resource.ResourceType;
+import net.minecraft.resource.ZipResourcePack;
 import net.minecraft.util.Identifier;
 
+/**
+ * Make the default resource pack use the MC jar directly instead of the full classpath.
+ * This is a major speed improvement, as well as a bugfix (it prevents other mod jars from overriding MC's resources).
+ */
 @Mixin(DefaultResourcePack.class)
-public class DefaultResourcePackMixin {
-	@Inject(method = "findInputStream", at = @At("HEAD"), cancellable = true)
-	protected void onFindInputStream(ResourceType resourceType, Identifier identifier, CallbackInfoReturnable<InputStream> callback) {
-		if (DefaultResourcePack.resourcePath != null) {
-			// Fall through to Vanilla logic, they have a special case here.
-			return;
-		}
+public abstract class DefaultResourcePackMixin {
+	/**
+	 * Redirect all resource access to the MC jar zip pack.
+	 */
+	final ResourcePack fabric_mcJarPack = createJarZipPack();
```

Comment:
Minor thing, no need to change now. But you can use @Unique on the field without the fabric_ prefix.